_posts/2025-05-05-plan-execute-agent.md

```diff
@@ -0,0 +1,423 @@
+        "resource.attributes.service@version": "2.0.2",
+        "resource.attributes.telemetry@sdk@name": "opentelemetry",
+        "resource.attributes.telemetry@sdk@language": "dotnet",
+        "resource.attributes.host@name": "7cbfd02de5b9",
+        "resource.attributes.service@namespace": "opentelemetry-demo",
+        "resource.attributes.container@id": "7cbfd02de5b9886dcda65e9876ee4d9601c98b29efada503117f3ac187a3b81a",
+        "resource.attributes.telemetry@sdk@version": "1.11.2",
+        "resource.attributes.service@name": "cart",
+        "log.attributes.StatusCode": "FailedPrecondition"
+    }
+}
+```
+
+As seen, the root cause clearly points to a Redis connection issue. 
+
+For more details on agent registration and execution, see this [tutorial](https://docs.opensearch.org/docs/latest/tutorials/gen-ai/agents/build-plan-execute-reflect-agent/).
+
+If you wish to look at intermediate results, you can query the memory index and memory traces to understand the interactions between the agent and the LLM. 
+
+Fetch message traces
+```json
+GET /_plugins/_ml/memory/message/your_message_id/traces
+```
+For more details about memory and traces, see [Memory](https://docs.opensearch.org/docs/latest/ml-commons-plugin/api/memory-apis/index/).
+
+## Extending the Agent with More Tools and MCP Support
+
+The power of the Plan–Execute–And-Reflect agent lies in its flexibility. While our example uses a basic set of tools to investigate a cart failure, the agent can support a wide range of tools. For the list of available tools, see [Tools](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/tools/index/).
+
+One such example would be to use the [WebSearchTool](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/tools/web-search-tool/) to perform deep-research tasks. 
+
+In addition, this agent acts as an MCP client, meaning it can connect to an MCP (Model Coordination Protocol) server. This allows the agent to retrieve tools and configurations dynamically and participate in more complex workflows.
+
+If you’d like to connect your agent to an MCP server, refer to this [document](https://docs.opensearch.org/docs/latest/ml-commons-plugin/agents-tools/mcp/mcp-connector/) for detailed instructions.
+
+### Recommendations and troubleshooting
+
+* If you notice your agent getting throttled by your model, enable retries via the [connector configuration](https://docs.opensearch.org/docs/latest/ml-commons-plugin/remote-models/blueprints/#configuration-parameters).
+* If the agent stops with max_steps executed, then you can increase the [max_steps](https://docs.opensearch.org/docs/latest/ml-commons-plugin/api/agent-apis/register-agent/#request-body-fields) or specify it during execution via the `parameters.max_steps` field.
+
+Note: This is an experimental feature and will continue to evolve with future enhancements.
+
+### Future Enhancements
```

Comment:
I think no need to mention this. As we already mentioned check details in Github issue 

> For updates on the progress of the feature or if you want to leave feedback, see the associated GitHub [issue](https://github.com/opensearch-project/ml-commons/issues/3745).